### ISSUE TITLE
sinon@next: Shouldn't fake.yields() have the same semantics as stub().yields()?

### ISSUE THREAD

**Issue by mantoni** [sinonjs/sinon#1697]
The current implementation of `fake.yields()` is completely different from what `yields` meant for stubs. Is this intentional @mroderick? I would prefer keeping it in line with the `stub` semantics so that a migration from stubs to fakes is not becoming too confusing.

I think what `fake.yields(callback)` is currently doing should really be named `fake.calls(callback)`.

Via #1586

**Comment by mroderick** [comment]
> The current implementation of fake.yields() is completely different from what yields meant for stubs. Is this intentional? I would prefer keeping it in line with the stub semantics so that a migration from stubs to fakes is not becoming too confusing.

I agree. Let's try to make the transition as easy as possible.

> I think what `fake.yields(callback)` is currently doing should really be named `fake.calls(callback)`.

That would be the same as `fake(callback)` or `sinon.spy(callback)`, wouldn't it? If that's correct, then we don't need to add it.

**Comment by mantoni** [comment]
Yes, you're right. We actually don't need "calls". I just re-discovered that there is another variation of `yields`: `yieldsRight`. I remember being confused that `yields` would invoke the first function it finds in the argument list. `yieldsRight` is what would have been intuitive to me.

I would suggest to make `yields` very strict and attempt to invoke the last argument as a function. If it's not a function it should throw. If someone wants to invoke a different argument as a function, they can do so with `callsArg(index, args...)`.

**Comment by mroderick** [comment]
> I would suggest to make `yields` very strict and attempt to invoke the last argument as a function. If it's not a function it should throw

That sounds like a reasonable approach to me 👍 

> If someone wants to invoke a different argument as a function, they can do so with callsArg(index, args...).

When creating a fake, these properties were removed (as they contribute to make it stateful, or are just adding noise)

* `instantiateFake`
* `callArg`
* `callArgOn`
* `callArgOnWith`
* `callArgWith`
* `invokeCallback`
* `throwArg`
* `withArgs`
* `yield`
* `yieldOn`
* `yieldTo`
* `yieldToOn`

I think `fake` is capable enough as it is, and would like to wait with adding more features, before we know we have to.

**Comment by mantoni** [comment]
I'm quite sure that we'll have to add some of those functions back in. I've already experienced cases where I wanted `yield` myself.

I would also like to emphasise the differences between `yield` and `yields` (or `callArg` and `callsArg`). The first invoked the callback immediately while the latter programs the fake to always invoke the callback. Both variations have their use cases and I would clearly not be able to migrate tests if one of the two versions is missing.

Having said that, I think the issues we faced in the past had their origin in cases where behaviour was mixed with `withArgs` calls. That is clearly to avoid in the fakes implementation. I'd also drop less memorable convenience methods like `callArgOnWith`. It's almost less typing to simply wrap a custom implementation 😆 

Anyway, to keep things focused, I'd suggest to move forward with care and discuss each of those topics in separate issues. It's quite a large API that we have here. As for this issue, I'm going to send a PR that adjusts the `yields` and `yieldsAsync` behaviour and nothing else.
